src/server/pfs/server/driver_compact.go

```diff
@@ -14,48 +14,63 @@ import (
 	"golang.org/x/net/context"
 )
 
-func (d *driver) compact(master *work.Master, ids []fileset.ID) (*fileset.ID, error) {
-	workerFunc := func(ctx context.Context, tasks []fileset.CompactionTask) ([]fileset.ID, error) {
-		workTasks := make([]*work.Task, len(tasks))
-		for i, task := range tasks {
-			serInputs := make([]string, len(task.Inputs))
-			for i := range task.Inputs {
-				serInputs[i] = task.Inputs[i].HexString()
+func (d *driver) compact(ctx context.Context, ids []fileset.ID) (*fileset.ID, error) {
+	compacted, err := d.storage.IsCompacted(ctx, ids)
+	if err != nil {
+		return nil, err
+	}
+	if compacted {
+		return d.storage.Compose(ctx, ids, defaultTTL)
+	}
+	var id *fileset.ID
+	if err := d.compactionQueue.RunTaskBlock(ctx, func(master *work.Master) error {
```

Comment:
I do like this better here instead of in finish commit. But I think not having will create concurrency issues in finishCommit. See my other comment.

Reply:
This doesn't lock anything. It just sets up the task in the work package and blocks until it is complete. We need to solve the concurrency issue through Postgres as mentioned above.